pkg/infrastructure/clusterapi/clusterapi.go

```diff
@@ -0,0 +1,328 @@
+package clusterapi
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"gopkg.in/yaml.v2"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	utilkubeconfig "sigs.k8s.io/cluster-api/util/kubeconfig"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager/signals"
+
+	"github.com/openshift/installer/pkg/asset"
+	"github.com/openshift/installer/pkg/asset/cluster/metadata"
+	"github.com/openshift/installer/pkg/asset/ignition/bootstrap"
+	"github.com/openshift/installer/pkg/asset/ignition/machine"
+	"github.com/openshift/installer/pkg/asset/installconfig"
+	"github.com/openshift/installer/pkg/asset/kubeconfig"
+	"github.com/openshift/installer/pkg/asset/manifests/capiutils"
+	capimanifests "github.com/openshift/installer/pkg/asset/manifests/clusterapi"
+	"github.com/openshift/installer/pkg/clusterapi"
+	"github.com/openshift/installer/pkg/infrastructure"
+	"github.com/openshift/installer/pkg/types"
+)
+
+// InfraProvider implements the infrastructure/provider interface
+// which is the interface between the rest of the installer and
+// the CAPI infrastructure provider.
+type InfraProvider struct {
+	infrastructure.Provider
+
+	capiProvider Provider
+}
+
+// InitializeProvider returns a CAPI provider implementation for a specific
+// cloud platform.
+func InitializeProvider(platform Provider) infrastructure.Provider {
+	return InfraProvider{capiProvider: platform}
+}
+
+// Provider provides an interface, intended for platform-specific implementations,
+// for calling functions at different points of the CAPI infrastructure provisioning
+// lifecycle.
+type Provider interface {
+	// PreProvision is called before provisioning using CAPI controllers has begun.
+	// and should be used to create dependencies needed for CAPI provisioning,
+	// such as IAM roles or policies.
+	PreProvision(in PreProvisionInput) error
+
+	// Ignition generates the ignition secrets for bootstrap and control-plane machines
+	// and handles any preconditions for ignition.
+	Ignition(in IgnitionInput) ([]client.Object, error)
+
+	// ControlPlaneAvailable is called once cluster.Spec.ControlPlaneEndpoint.IsValid()
+	// returns true, typically after load balancers have been provisioned. It can be used
+	// to create DNS records.
```

Comment:
🚨This interface does not work for OpenStack🚨

For OpenStack we will create a floating IP for the API loadbalancer *before* creating the CAPI manifests. We will populate `ControlPlaneEndpoint` in the spec when generating it. Therefore, for OpenStack, this hook would be called immediately the cluster object is created. However, the infrastructure is not ready at this point.

This field is normally provisioned by the cluster controller when the API loadbalancer is ready. However, CAPO cannot currently provision the type of LoadBalancer used by Shift on Stack. Consequently we disable all loadbalancer provisioning in CAPO, so nothing will ever populate this field. We need the hook to be called when the cluster infrastructure is provisioned so we can connect the Floating IP we pre-created to the network that CAPO provisioned.

What we need is available in the InfrastructureReady condition on the cluster object, documented here: https://cluster-api.sigs.k8s.io/developer/architecture/controllers/cluster#infrastructure-provider

When the infrastructure cluster controller (CAPO) has provisioned its infrastructure it sets `status.Ready`. When the cluster controller (CAPI) observers *both* `status.Ready` and that `ControlPlaneEndpoint` is valid it will set the `InfrastructureReady` condition on the cluster object. We need the hook to be called when `InfrastructureReady` is set.

Reply:
I would also like to consider adding an _additional_ hook for when Infrastructure is ready. It seems like `ControlPlaneEndpointAvailable` and `InfrastructureReady` could coexist.

Reply:
Note that with this API, ControlPlaneEndpointAvailable will be called on an OpenStack cluster before the control plane endpoint is available. Do any providers have a reason to call this if the cluster infrastructure is not ready? I suspect that just switching this to use InfrastructureReady instead would be a no-op (near enough) on other providers.

Reply:
Discussed [in Slack](https://redhat-internal.slack.com/archives/C05KZA3NVU6/p1705509841892389), but summarised here:

CAPO cannot provision the kind of loadbalancer that OpenShift expects. We know what the endpoint will be when we generate the manifests, but nothing in CAPO can configure that endpoint in the required way. We can only hook it up correctly once the cluster has provisioned the cluster network. Consequently we just disable all API loadbalancer provisioning in the CAPO cluster spec and populate ControlPlaneEndpoint directly. CAPI can use ControlPlaneEndpoint in the normal manner and it will work correctly after the callback has run.

Incidentally, CAPO can configure an API loadbalancer in 2 different ways, both of which will populate ControlPlaneEndpoint automatically, but OpenShift doesn't support either of them. Resolving this is future work.

Reply:
We now check InfrastructureReady. Thanks for catching this!